broker-daemon/state-machines/interchain.js

```diff
@@ -0,0 +1,292 @@
+ *                                          expected inbound payment.
+ * @param {Date}    timeout               - Absolute time after which the
+ *                                          payment should not be translated.
+ */
+async function prepareSwap (hash, { engine, amount }, timeout) {
+  // TODO: update `prepareSwap` to take an expiration
+  // time (absolute) and a minimum time lock in seconds
+  // TODO: update `prepareSwap` to be idempotent
+  await engine.prepareSwap(
+    hash,
+    amount,
+    INBOUND_TIME_LOCK,
+    timeout
+  )
+}
+
+/**
+ * The result of retrieving a preimage by paying a downstream payment locked
+ * with the swap hash.
+ * @typedef {Object} PaymentResult
+ * @property {string} paymentPreimage - Base64 encoded preimage for the swap.
+ * @property {string} paymentError    - Error message from a failed attempt to
+ *                                      translate a swap via payment.
+ *
+ */
+
+/**
+ * Get the preimage for a swap from one of its three potential sources:
+ * - from an existing outbound payment
+ * - from an existing settled inbound invoice
+ * - from a new outbound payment
+ *
+ * @private
+ *
+ * @param {string}  hash                    - Base64 string of the hash for the
+ *                                            swap
+ * @param {Payment} inboundPayment
+ * @param {Engine}  inboundPayment.engine   - Engine of the expected inbound
+ *                                            payment
+ * @param {Payment} outboundPayment
+ * @param {Engine}  outboundPayment.engine  - Engine of the outbound
+ *                                            payment
+ * @param {number}  outboundPayment.amount  - Amount, in the smallest unit, of
+ *                                            the outbound payment to be made.
+ * @param {number}  outboundPayment.address - Address to send the outbound
+ *                                            payment to to retrieve the
+ *                                            preimage.
+ * @returns {PaymentResult}                   Result of retrieving the preimage.
+ */
+async function getPreimage (
+  hash,
+  { engine: inboundEngine },
+  { engine: outboundEngine, amount: outboundAmount, address: outboundAddress }
+) {
+  // We don't want to reject an incoming HTLC if we know that there is an
+  // active outgoing one. If the outgoing HTLC is in flight or completed, we
+  // should attempt to retrieve the associated preimage.
+  logger.debug(`Checking outbound HTLC status for swap ${hash}`)
+  if (await outboundEngine.isPaymentPendingOrComplete(hash)) {
+    // TODO: return permanent errors when encountered
+    return {
+      paymentPreimage: await outboundEngine.getPaymentPreimage(hash)
+    }
+  }
+
+  // wait for an incoming payment to this hash that is accepted,
+  // but not yet settled.
+  try {
+    // TODO: implement this method
+    await inboundEngine.waitForSwapCommitment(hash)
+  } catch (e) {
+    // TODO: fix this error message
+    if (e.isSettled) {
```

Comment:
@clark800 how do I get access to the custom error types exposed in https://github.com/sparkswap/lnd-engine/pull/234 ? I think I need them to replace my current check.

Reply:
They are exported here: https://github.com/sparkswap/lnd-engine/blob/master/src/engine-actions/index.js#L96